### PR TITLE
[Test] Add extra test for val_check_interval in distributed scenario

### LINKED ISSUE #7751: Progress bar inaccurate during multi-node training [Fairscale/Deepspeed]
## 🐛 Bug

When using either Fairscale or Deepspeed on multiple gpus across multiple nodes, the progress bar seems to be inaccurate. The total iterations seems to be calculated using the per gpu batch size. At first I thought, the current iteration count may be not matching up if it was based on the effective batch size (per_gpu_batch_size * num_gpus * num_nodes * accumulation_steps), however that doesn't perfectly match up either. I've looked through the documentation, but it doesn't really seem to specify weather the intent with the progress bar was to use the per gpu batch size or the effective bath size across all gpus. I have tried explicitly setting  `logging_batch_size_per_gpu` in the Deepspeed plugin as well to no avail, and same thing happens with Fairscale based training as well.

### To Reproduce

```python
trainer_args = dict()
trainer_args["plugins"] = []
trainer_args["num_nodes"] = 4
trainer_args["gpus"] = 4
trainer_args["accelerator"] = "ddp"
trainer_args["logger"] = mlf_logger
trainer_args["default_root_dir"] = PATH_LOG
trainer_args["callbacks"] = [model_checkpoint]

trainer_args["max_epochs"] = 1 
trainer_args["val_check_interval"] = 0.05  
trainer_args["accumulate_grad_batches"] = 1

train_mode = DeepSpeedPlugin(
    logging_batch_size_per_gpu = 128,
    cpu_offload = True,
    allgather_bucket_size=3e8,
    reduce_bucket_size=3e8,
)
trainer_args["plugins"].append(train_mode)
trainer_args["precision"] = 16

data_module = experiment_data.DataDefinitionA(
    batch_size=128,
    num_workers=6,
    prefetch_factor=10,
)

optimizer_config = {
    "optimizer_class": DeepSpeedCPUAdam, #FusedAdam, 
    "optimizer_kwargs": {"lr" : 5e-5},
}

model = experiment_models.bert_config_A(
    tokens_to_indices=data_module.vocab_and_mappings["tokens_to_indices"],
    config_positional=data_module.vocab_and_mappings["config_positional"],
    max_len=data_module.max_len,
    optimizer_config=optimizer_config,
    use_sent_idx=True,
    which='SMALL'
)

trainer = Trainer(**trainer_args)

trainer.fit(model, data_module)
```

Log at the end of training:
Note above it is set to train for 1 epoch, yet at the end of training the progress bar is only 3%.
15974338 total iterations seems correct assuming its based on the per gpu batch size like `total iterations = (num_training_samples + num_val_samples/val_check_interval) / batch_size_per_gpu` However, 528947 * 16 gpus != 15974338 so I am really quite unsure of how the current iteration is being set.
```
> Epoch 0:   3%|▎         | 528946/15974338 [145:20:08<4243:51:07,  1.01it/s, loss=1.96, v_num=61fd]
> Epoch 0:   3%|▎         | 528946/15974338 [145:20:08<4243:51:07,  1.01it/s, loss=1.97, v_num=61fd]
> Epoch 0:   3%|▎         | 528947/15974338 [145:20:09<4243:50:51,  1.01it/s, loss=1.97, v_num=61fd]
> Epoch 0:   3%|▎         | 528947/15974338 [145:20:09<4243:50:51,  1.01it/s, loss=1.97, v_num=61fd]
> Epoch 0:   3%|▎         | 528947/15974338 [145:22:22<4244:55:27,  1.01it/s, loss=1.97, v_num=61fd]
> 
> 
> [2021-05-23T10:36:42.272833] The experiment completed successfully. Finalizing run...
> 2021-05-23 10:36:42,278 - __main__ - 124 - INFO - Exiting context: TrackUserError
> 2021-05-23 10:36:42,278 - __main__ - 124 - INFO - Exiting context: RunHistory
> Cleaning up all outstanding Run operations, waiting 900.0 seconds
> 2 items cleaning up...
> Cleanup took 0.14957070350646973 seconds
> 2021-05-23 10:36:42,582 - __main__ - 124 - INFO - Exiting context: Dataset
> 2021-05-23 10:36:42,582 - __main__ - 124 - INFO - Exiting context: ProjectPythonPath
> [2021-05-23T10:36:42.583066] Finished context manager injector.
```

### Expected behavior
Progress bar reaches 100% at the end of the epoch and correctly updates iterations through training so the training time estimate is also accurate. 

### Environment

 - PyTorch Version (e.g., 1.0): 1.8.1
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): N/A
 - Python version: 3.8
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: 16 v100s across 4 nodes
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

From some investigation, the current behaviour seems correct. 

When using val_check_interval. If you have 8 train batches, and val_check_interval is 0.2, val_check_batch=1 as `val_check_batch=max(1, int(train_batches * val_check_interval))`.

Alternatively, we could randomly sample int(1 / val_check_interval) batches from train to run the validation on, but the validation won't be run at the frequency, but it might be confusing for the users.

This PR adds a test to make sure the correct number total number of batches is selected.

Additionally, it adds a section in the doc to compute the total number of batches.

Fixes #7751

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by SeanNaren — APPROVED** [review]
Nice check!

File: tests/trainer/test_dataloaders.py [hunk 1]
```diff
@@ -1665,3 +1665,47 @@ def on_validation_batch_start(self, batch, batch_idx: int, dataloader_idx: int)
...
+    "total_train_samples",
+    "train_batch_size",
+    "total_val_samples",
+    "val_batch_size",
+    "val_check_interval",
+], [
+    pytest.param(8, 4, 2, 1, 0.2),
+    pytest.param(8, 4, 2, 1, 0.5),
+])
+def test_progress_bar_max_val_check_interval(
```
Comment (carmocca) [inline tests/trainer/test_dataloaders.py:1681]: I think this test would fit better in the train loop tests file

File: tests/trainer/test_dataloaders.py [hunk 2]
```diff
@@ -1665,3 +1665,47 @@ def on_validation_batch_start(self, batch, batch_idx: int, dataloader_idx: int)
...
+):
+
+    world_size = 2
+
+    train_data = DataLoader(RandomDataset(32, total_train_samples), batch_size=train_batch_size)
+    val_data = DataLoader(RandomDataset(32, total_val_samples), batch_size=val_batch_size)
+
+    model = BoringModel()
+    trainer = Trainer(
+        default_root_dir=os.getcwd(),
```
[the remark this answers is not in the document]
Comment (tchaton) [reply tests/trainer/test_dataloaders.py]: Changed for `tmpdir`.

File: tests/trainer/test_dataloaders.py [hunk 3]
```diff
@@ -1665,3 +1665,47 @@ def on_validation_batch_start(self, batch, batch_idx: int, dataloader_idx: int)
...
+@RunIf(min_gpus=2, special=True)
+@pytest.mark.parametrize([
+    "total_train_samples",
+    "train_batch_size",
+    "total_val_samples",
+    "val_batch_size",
+    "val_check_interval",
+], [
+    pytest.param(8, 4, 2, 1, 0.2),
+    pytest.param(8, 4, 2, 1, 0.5),
```
[the remark this answers is not in the document]
Comment (tchaton) [reply tests/trainer/test_dataloaders.py:1679]: Done !

File: tests/trainer/test_dataloaders.py (hunk 1 above)
Comment (tchaton) [reply tests/trainer/test_dataloaders.py:1681]: Actually, it would be better in progress `test_progress_bar.py`

**Review by carmocca — APPROVED** [review]

File: tests/callbacks/test_progress_bar.py [hunk 4]
```diff
@@ -533,3 +535,47 @@ def test_progress_bar_can_be_pickled():
...
+    total_train_batches = total_train_samples // (train_batch_size * world_size)
+    val_check_batch = max(1, int(total_train_batches * val_check_interval))
+    assert trainer.val_check_batch == val_check_batch
+    val_checks_per_epoch = total_train_batches / val_check_batch
+    total_val_batches = total_val_samples // (val_batch_size * world_size)
+    assert trainer.callbacks[0].total_train_batches == total_train_batches
+    assert trainer.callbacks[0].total_val_batches == total_val_batches
+    total_val_batches = total_val_batches * val_checks_per_epoch
+    if trainer.is_global_zero:
+        assert trainer.callbacks[0].main_progress_bar.total == total_train_batches + total_val_batches
```
Comment (awaelchli) [inline tests/callbacks/test_progress_bar.py:581]:
```suggestion
    assert trainer.progress_bar_callback.total_train_batches == total_train_batches
    assert trainer.progress_bar_callback.total_val_batches == total_val_batches
    total_val_batches = total_val_batches * val_checks_per_epoch
    if trainer.is_global_zero:
        assert trainer.progress_bar_callback.main_progress_bar.total == total_train_batches + total_val_batches
```

**Review by awaelchli — APPROVED** [review]
